finetune/nn/target_blocks.py

```diff
@@ -65,9 +65,76 @@ def cps_model(M, cps_mask, hidden, config):
     return {"loss": mlm_loss, "logits": logits}
 
+def masked_language_model_(*, X, M, mlm_weights, mlm_ids, mlm_positions, targets, embed_weights, hidden, config, reuse=None, train=False):
+    with tf.variable_scope('model/masked-language-model'):
+        batch, seq, feats = shape_list(hidden)
+        flat_offsets = tf.reshape(
+            tf.range(0, batch, dtype=tf.int32) * seq, [-1, 1]
+        )
+
+        not_padding = tf.reshape(mlm_weights, [-1]) > 1e-9
+        flat_positions = tf.boolean_mask(tf.reshape(mlm_positions + flat_offsets, [-1]), not_padding) # take off the padding entirely
+        gathered_hidden = tf.gather(tf.reshape(hidden, [batch * seq, feats]), flat_positions)
+        # Commenting this out and replacing with simple positional loss
+        mlm_ids = tf.boolean_mask(tf.reshape(mlm_ids, [-1]), not_padding)
+        # Unsure if this is directly applicable--dimension should be (batch, seq, 1))
+        # as far as I understand it, the not_padding stuff is generated from mlm_weights
+        # which should represent the mask.
+        targets = tf.boolean_mask(tf.reshape(targets, [-1,4]), not_padding)
+        final_proj = dense_with_custom_init(
+            gathered_hidden,
+            # Projection down to a single value for regression
+            config.n_embed,
+            activation=act_fns[config.act_fn],
+            kernel_initializer=tf.random_normal_initializer(stddev=config.weight_stddev),
+            name='dense',
+            custom=config.use_auxiliary_info and not (config.mlm_baseline or config.pos_injection),
+            pos_embed=config.n_context_embed_per_channel * config.context_dim,
+            proj_type='downward',
+            transpose_b=True
+        )
+        normed_proj = norm(final_proj, 'LayerNorm')
+        n_vocab = shape_list(embed_weights)[0]
+        output_bias = tf.get_variable(
+            #"output_bias",
+            "output_bias_posreg",
+            #shape=[n_vocab],
+            # We are regressing into 4D space...
+            # might be a good idea to do top left + offset?
+            shape=[4],
+            initializer=tf.zeros_initializer()
+        )
+        
+        # Going to add a new variable to project downwards
+        regression_matrix = tf.get_variable(
+                "regression_matrix",
+                shape=[n_vocab, 4],
+                initializer=tf.zeros_initializer(),
+                dtype=tf.float32
+        )
+
+        logits = tf.matmul(normed_proj, embed_weights, transpose_b=True)
+        logits = tf.reshape(tf.matmul(logits, regression_matrix), [-1,4])
+        logits = tf.nn.bias_add(logits, output_bias)
+        # Swapping out old logit stuff
+        #mlm_loss = tf.contrib.losses.sparse_softmax_cross_entropy(
+        mlm_loss = tf.contrib.losses.mean_squared_error(
+            #logits,
+            tf.broadcast_to(output_bias, tf.shape(targets)),
```

Comment:
Only comment here is that I don't think you get a lot of value out of running this experiment that you wouldn't get out of purely computing the mean and calculating MSE offline. The point in my mind was just to get some sort of baseline that we could use to understand how well the model is performing. 

As I think about this more, when you continue these experiments it would be nice to log something like mean error per dim or something that would give us a quick way to get a qualitative sense of how well the model is doing.


I.e.:

```
mean_error = tf.math.abs(tf.mean(logits - targets, axis=0))
tf.summary.scalar("Top", mean_error[0])
tf.summary.scalar("Bottom", mean_error[1])
...
```

Reply:
That seems like a reasonable thing to do. I might actually want to try absolute error, also!

Reply:
yeah i think only argument against MAE is that you have a sharp discontinuity at 0, but that's only relevant late in training